File: day03/aoc3-2.py
def largest_bank_number_twelve(line: str, k: int = 12) -> int:
    digits = [int(ch) for ch in line if ch.isdigit()]
    n = len(digits)
    if n < k:
        return -1

    removals = n - k
    stack = []

    for d in digits:
        while removals > 0 and stack and stack[-1] < d:
            stack.pop()
            removals -= 1
        stack.append(d)

    # Trim if still too long
    if len(stack) > k:
        stack = stack[:k]

    return int("".join(str(x) for x in stack))


def process_file_twelve(filename: str = "aocdata.txt"):
    total = 0
    with open(filename) as f:
        for line in f:
            s = line.strip()
            if not s:
                continue
            val = largest_bank_number_twelve(s, 12)
            print(f"Bank: {s} -> Largest possible 12-digit number: {val}")
            if val != -1:
                total += val
    print(f"Total sum of largest 12-digit numbers: {total}")


# Run with test data
process_file_twelve("aocdata.txt")

File: day03/aoc3-2.pl
#!/usr/bin/perl
use strict;
use warnings;

sub largest_bank_number {
    my ($line, $k) = @_;   # $k = length of number to build (12)
    my @digits = grep { /^\d$/ } split //, $line;
    my $n = scalar(@digits);

    return -1 if $n < $k;  # not enough digits

    my @result;
    my $to_pick = $k;
    my $start = 0;

    # Greedy selection: at each step, pick the largest digit possible
    while ($to_pick > 0) {
        # Remaining digits we must leave room for
        my $end = $n - $to_pick;
        my $max_digit = -1;
        my $pos = $start;

        # Find the largest digit we can pick within the allowed window
        for my $i ($start .. $end) {
            if ($digits[$i] > $max_digit) {
                $max_digit = $digits[$i];
                $pos = $i;
            }
        }

        push @result, $max_digit;
        $start = $pos + 1;
        $to_pick--;
    }

    return join("", @result);
}

# Read from aocdata.txt
my $filename = "aocdata.txt";
open my $fh, '<', $filename or die "Could not open '$filename': $!";

my $sum = 0;

while (my $line = <$fh>) {
    chomp $line;
    next if $line =~ /^\s*$/;  # skip blank lines
    my $largest = largest_bank_number($line, 12);
    print "Bank: $line -> Largest possible 12-digit number: $largest\n" if $largest ne "-1";
    $sum += $largest if $largest ne "-1";
}

close $fh;

print "Total sum of largest 12-digit numbers: $sum\n";
